bpf/kmesh/ads/include/local_ratelimit.h

```diff
@@ -0,0 +1,201 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Kmesh */
+
+#ifndef __KMESH_LOCAL_RATE_LIMIT_H__
+#define __KMESH_LOCAL_RATE_LIMIT_H__
+
+#include "bpf_log.h"
+#include "bpf_common.h"
+#include "kmesh_common.h"
+#include "listener/listener.pb-c.h"
+
+struct ratelimit_key {
+    union {
+        struct {
+            __u64 netns;  /* Network namespace. */
+            __u32 ipv4;   /* Destination IPv4 address. */
+            __u32 port;   /* Destination port. */
+            __u32 family; /* Address family (e.g., AF_INET) */
+        } sk_skb;
+    } key;
+};
+
+struct ratelimit_value {
+    __u64 last_topup; /* Timestamp of the last token refill (nanoseconds) */
+    __u64 tokens;     /* Current number of available tokens */
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __type(key, struct ratelimit_key);
+    __type(value, struct ratelimit_value);
+    __uint(max_entries, MAP_SIZE_OF_LISTENER);
+    __uint(map_flags, BPF_F_NO_PREALLOC);
+} kmesh_ratelimit SEC(".maps");
+
+struct ratelimit_settings {
+    __u64 max_tokens;      /* Maximum capacity of the token bucket */
+    __u64 tokens_per_fill; /* Number of tokens added per refill */
+    __u64 fill_interval;   /* Interval between token refills (nanoseconds) */
+};
+
+/**
+ * Enforces rate limiting for a given key using token bucket algorithm.
+ *
+ * @param key       Pointer to the rate limit key (e.g., IP and port).
+ * @param settings  Pointer to rate limit settings (bucket size, refill rate, interval).
+ * @return          0 if allowed (token consumed), -1 if rate limit exceeded.
+ */
+static inline int rate_limit__check_and_take(struct ratelimit_key *key, const struct ratelimit_settings *settings)
+{
+    struct ratelimit_value *value;
+    struct ratelimit_value new_value;
+    __u64 now = bpf_ktime_get_ns();
+    __u64 topup;
+
+    value = bpf_map_lookup_elem(&kmesh_ratelimit, key);
+    if (!value) {
+        new_value.last_topup = now;
+        new_value.tokens = settings->max_tokens;
+        bpf_map_update_elem(&kmesh_ratelimit, key, &new_value, BPF_NOEXIST);
+        value = bpf_map_lookup_elem(&kmesh_ratelimit, key);
+        if (!value)
+            return 0;
+    }
+
+    topup = (now - value->last_topup) / settings->fill_interval;
+    if (topup > 0) {
+        value->tokens += topup * settings->tokens_per_fill;
+        if (value->tokens > settings->max_tokens) {
+            value->tokens = settings->max_tokens;
```

Comment:
This is not concurrent safe

Reply:
can this ensue concurent safe ? cause only one progress can add the value.

```c
  topup = (now - value->last_topup) / settings->fill_interval;
  if (topup > 0) {
      topup_time = value->last_topup + topup * settings->fill_interval;
      if (__sync_bool_compare_and_swap(&value->last_topup, value->last_topup, topup_time)) {
          delta = topup * settings->tokens_per_fill;
          if (value->tokens + delta > settings->max_tokens) {
              delta = settings->max_tokens - value->tokens;
          }
          __sync_fetch_and_add(&value->tokens, delta);
      }
  }

```

Reply:
@nlgwcy is an expert on this